### ISSUE TITLE
image error  with Fine-tuning RRDB using SSL-loss

### ISSUE THREAD

**Issue by ningbende** [ChrisDud0257/SSL#6]
Hi ChrisDud0257,

Thank you for your open source project! I encountered some issues while fine-tuning the RRDB model using SSL-loss. Specifically, the results from the model show a lot of artifacts, as shown in the attached image:

 
![Snipaste_2024-10-25_10-17-15](https://github.com/user-attachments/assets/117655f6-1978-43bc-bb3f-6da4c29aecaa)

Have you experienced similar issues? If so, how did you address them?

Thank you for your help!

**Comment by ChrisDud0257** [comment]
> Hi ChrisDud0257,
> 
> Thank you for your open source project! I encountered some issues while fine-tuning the RRDB model using SSL-loss. Specifically, the results from the model show a lot of artifacts, as shown in the attached image:
> 
> ![Snipaste_2024-10-25_10-17-15](https://private-user-images.githubusercontent.com/4524235/379988686-117655f6-1978-43bc-bb3f-6da4c29aecaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk4NDA1NzgsIm5iZiI6MTcyOTg0MDI3OCwicGF0aCI6Ii80NTI0MjM1LzM3OTk4ODY4Ni0xMTc2NTVmNi0xOTc4LTQzYmMtYmIzZi02ZGE0YzI5YWVjYWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjVUMDcxMTE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTU1ZTllYmUyNGYzNWNiN2QxNmQ1ZTdiZTRlMTkwMmJlZjRhZjUyN2Y1NmUwMTAxYzRkMWIzODA2NDc3NzlmMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.T89tm-R6qhFBJFw5pAn1njAidAMVkkQ_8rHxp4HdJzY)
> 
> Have you experienced similar issues? If so, how did you address them?
> 
> Thank you for your help!

Have you utilized SSL under GAN or diffusion models? If no generative loss is included, then the SSL loss will have some side effect on the model, we also don't  recommend this training manner. Or if you have tried it under generative models,  perhaps you have set the weight of SSL to a too large value. Please decrease it. The magnitude of SSL should not surpass the largest loss term in your implementation.

**Comment by ningbende** [comment]
I am currently training the network using the default parameters provided in the ESRGANSSLModel as outlined in the paper. I am preparing for a 2x super-resolution training with the following configuration:
######################
```python
ssl_setting:
  mask_stride: 3
  ssl_mode: 'cuda'  # You could change it to "pytorch" to see the detailed implementation of our SSL. But we don't recommend you to do that since the pytorch version will need at least 48GB GPU memory
  kernel_size_search: 25
  sigma: 0.004  # scaling_factor in the paper
  kernel_size_window: 9
  generalization: True

# training settings
train:
  ema_decay: 0.999
  optim_g:
    type: Adam
    lr: !!float 1e-4
    weight_decay: 0
    betas: [0.9, 0.99]
  optim_d:
    type: Adam
    lr: !!float 1e-4
    weight_decay: 0
    betas: [0.9, 0.99]

  scheduler:
    type: MultiStepLR
    milestones: [50000, 100000, 200000, 300000]
    gamma: 0.5

  total_iter: 400000
  warmup_iter: -1  # no warm up

  # losses
  pixel_opt:
    type: L1Loss
    loss_weight: !!float 1e-2
    reduction: mean
  selfsim_opt:
    type: L1Loss
    loss_weight: !!float 1e3
    reduction: mean
  selfsim1_opt:
    type: KLDistanceLoss
    loss_weight: !!float 1e3
    reduction: mean
    softmax: False
  perceptual_opt:
    type: PerceptualLoss
    layer_weights:
      'conv5_4': 1  # before relu
    vgg_type: vgg19
    use_input_norm: true
    range_norm: false
    perceptual_weight: 1.0
    style_weight: 0
    criterion: l1
  gan_opt:
    type: GANLoss
    gan_type: vanilla
    real_label_val: 1.0
    fake_label_val: 0.0
    loss_weight: !!float 5e-3

  net_d_iters: 1
``` 

#######################
in the code about the SSL loss and its weight, I would like to know which parameters I should adjust to ensure a proper balance between the SSL component and other loss terms.

**Comment by ChrisDud0257** [comment]
> I am currently training the network using the default parameters provided in the ESRGANSSLModel as outlined in the paper. I am preparing for a 2x super-resolution training with the following configuration: ######################
> 
> ```python
> ssl_setting:
>   mask_stride: 3
>   ssl_mode: 'cuda'  # You could change it to "pytorch" to see the detailed implementation of our SSL. But we don't recommend you to do that since the pytorch version will need at least 48GB GPU memory
>   kernel_size_search: 25
>   sigma: 0.004  # scaling_factor in the paper
>   kernel_size_window: 9
>   generalization: True
> 
> # training settings
> train:
>   ema_decay: 0.999
>   optim_g:
>     type: Adam
>     lr: !!float 1e-4
>     weight_decay: 0
>     betas: [0.9, 0.99]
>   optim_d:
>     type: Adam
>     lr: !!float 1e-4
>     weight_decay: 0
>     betas: [0.9, 0.99]
> 
>   scheduler:
>     type: MultiStepLR
>     milestones: [50000, 100000, 200000, 300000]
>     gamma: 0.5
> 
>   total_iter: 400000
>   warmup_iter: -1  # no warm up
> 
>   # losses
>   pixel_opt:
>     type: L1Loss
>     loss_weight: !!float 1e-2
>     reduction: mean
>   selfsim_opt:
>     type: L1Loss
>     loss_weight: !!float 1e3
>     reduction: mean
>   selfsim1_opt:
>     type: KLDistanceLoss
>     loss_weight: !!float 1e3
>     reduction: mean
>     softmax: False
>   perceptual_opt:
>     type: PerceptualLoss
>     layer_weights:
>       'conv5_4': 1  # before relu
>     vgg_type: vgg19
>     use_input_norm: true
>     range_norm: false
>     perceptual_weight: 1.0
>     style_weight: 0
>     criterion: l1
>   gan_opt:
>     type: GANLoss
>     gan_type: vanilla
>     real_label_val: 1.0
>     fake_label_val: 0.0
>     loss_weight: !!float 5e-3
> 
>   net_d_iters: 1
> ```
> 
> ####################### in the code about the SSL loss and its weight, I would like to know which parameters I should adjust to ensure a proper balance between the SSL component and other loss terms.

I haven't trained a x2 super-resolution model before. Factually, we rarely do x2 GAN-based SR (expecially under bicubic degradation), since the x2 bicubic degradation is too weak to need a GAN loss to generate delicate details. On x4 ESRGAN-SSL, I haven't seen the artifacts which you have shown.

**Comment by ChrisDud0257** [comment]
> I am currently training the network using the default parameters provided in the ESRGANSSLModel as outlined in the paper. I am preparing for a 2x super-resolution training with the following configuration: ######################
> 
> ```python
> ssl_setting:
>   mask_stride: 3
>   ssl_mode: 'cuda'  # You could change it to "pytorch" to see the detailed implementation of our SSL. But we don't recommend you to do that since the pytorch version will need at least 48GB GPU memory
>   kernel_size_search: 25
>   sigma: 0.004  # scaling_factor in the paper
>   kernel_size_window: 9
>   generalization: True
> 
> # training settings
> train:
>   ema_decay: 0.999
>   optim_g:
>     type: Adam
>     lr: !!float 1e-4
>     weight_decay: 0
>     betas: [0.9, 0.99]
>   optim_d:
>     type: Adam
>     lr: !!float 1e-4
>     weight_decay: 0
>     betas: [0.9, 0.99]
> 
>   scheduler:
>     type: MultiStepLR
>     milestones: [50000, 100000, 200000, 300000]
>     gamma: 0.5
> 
>   total_iter: 400000
>   warmup_iter: -1  # no warm up
> 
>   # losses
>   pixel_opt:
>     type: L1Loss
>     loss_weight: !!float 1e-2
>     reduction: mean
>   selfsim_opt:
>     type: L1Loss
>     loss_weight: !!float 1e3
>     reduction: mean
>   selfsim1_opt:
>     type: KLDistanceLoss
>     loss_weight: !!float 1e3
>     reduction: mean
>     softmax: False
>   perceptual_opt:
>     type: PerceptualLoss
>     layer_weights:
>       'conv5_4': 1  # before relu
>     vgg_type: vgg19
>     use_input_norm: true
>     range_norm: false
>     perceptual_weight: 1.0
>     style_weight: 0
>     criterion: l1
>   gan_opt:
>     type: GANLoss
>     gan_type: vanilla
>     real_label_val: 1.0
>     fake_label_val: 0.0
>     loss_weight: !!float 5e-3
> 
>   net_d_iters: 1
> ```
> 
> ####################### in the code about the SSL loss and its weight, I would like to know which parameters I should adjust to ensure a proper balance between the SSL component and other loss terms.

Or one more thing, if you really need to try SSL under your setting, please increase the weight of perceptual loss weight to 10, and increase the GAN loss weight to 10. In the meantime, do not train the model for extended periods of time. Please observe the validation results during your training stage, once you think the validation images are just well, and the metrics are good, you could stop training.

**Comment by ChrisDud0257** [comment]
> > I am currently training the network using the default parameters provided in the ESRGANSSLModel as outlined in the paper. I am preparing for a 2x super-resolution training with the following configuration: ######################
> > ```python
> > ssl_setting:
> >   mask_stride: 3
> >   ssl_mode: 'cuda'  # You could change it to "pytorch" to see the detailed implementation of our SSL. But we don't recommend you to do that since the pytorch version will need at least 48GB GPU memory
> >   kernel_size_search: 25
> >   sigma: 0.004  # scaling_factor in the paper
> >   kernel_size_window: 9
> >   generalization: True
> > 
> > # training settings
> > train:
> >   ema_decay: 0.999
> >   optim_g:
> >     type: Adam
> >     lr: !!float 1e-4
> >     weight_decay: 0
> >     betas: [0.9, 0.99]
> >   optim_d:
> >     type: Adam
> >     lr: !!float 1e-4
> >     weight_decay: 0
> >     betas: [0.9, 0.99]
> > 
> >   scheduler:
> >     type: MultiStepLR
> >     milestones: [50000, 100000, 200000, 300000]
> >     gamma: 0.5
> > 
> >   total_iter: 400000
> >   warmup_iter: -1  # no warm up
> > 
> >   # losses
> >   pixel_opt:
> >     type: L1Loss
> >     loss_weight: !!float 1e-2
> >     reduction: mean
> >   selfsim_opt:
> >     type: L1Loss
> >     loss_weight: !!float 1e3
> >     reduction: mean
> >   selfsim1_opt:
> >     type: KLDistanceLoss
> >     loss_weight: !!float 1e3
> >     reduction: mean
> >     softmax: False
> >   perceptual_opt:
> >     type: PerceptualLoss
> >     layer_weights:
> >       'conv5_4': 1  # before relu
> >     vgg_type: vgg19
> >     use_input_norm: true
> >     range_norm: false
> >     perceptual_weight: 1.0
> >     style_weight: 0
> >     criterion: l1
> >   gan_opt:
> >     type: GANLoss
> >     gan_type: vanilla
> >     real_label_val: 1.0
> >     fake_label_val: 0.0
> >     loss_weight: !!float 5e-3
> > 
> >   net_d_iters: 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ####################### in the code about the SSL loss and its weight, I would like to know which parameters I should adjust to ensure a proper balance between the SSL component and other loss terms.
> 
> Or one more thing, if you really need to try SSL under your setting, please increase the weight of perceptual loss weight to 10, and increase the GAN loss weight to 10. In the meantime, do not train the model for extended periods of time. Please observe the validation results during your training stage, once you think the validation images are just well, and the metrics are good, you could stop training.

Sorry, I have made a typo, increase the GAN loss weight to 5e-2. You could also don't change the weights of other loss term as well, but decrease the weight of SSL to 500 or smaller. As have been illustrated in the end of this project, if you want to try SSL in your own task, maybe you should adjust some hyper-parameters according to your own requirements, to make the final performance just good.

**Comment by ningbende** [comment]
thanks so much , i will try
